### PR TITLE
[WIP] Bring Ammonite in closer alignment with argument conventions of other interpretters

### LINKED ISSUE #618: Ammonite parameter conventions out of alignment with common interpreters such as bash/Python/Ruby
In version 0.9.0, the argument passing behavior of Ammonite was changed in order to support script arguments directly following the script name. This particular functionality is compatible with other interpreters.

What is not expected, however, is:

- inability to provide ammonite parameters prior to the script in question
- the use of `--` is not consistent with other interpreters (script is the first argument following the `--`)

### PR DESCRIPTION
**Pull request by timcharper**
As documented in #618, it was not possible to specify ammonite arguments
if also specifying a scala script. With the change, the following
invocations work:

    amm --silent my-script.sc scriptarg1 scriptarg2
    amm --silent my-script.sc -- scriptarg1 scriptarg2
    amm my-script.sc --silent -- scriptarg1 scriptarg2
    amm my-script.sc --silent scriptarg1 scriptarg2

Additionally, we add assertions to prevent scriptargs from being
specified both before and after the delimiter. The following is
rejected:

    amm my-script.sc --silent scriptarg1 -- scriptarg2

Fixes #618

(we also add proper `--help`)

### CODE REVIEW

**Comment by timcharper** [comment]
Behavior is now as I would expect with this change:

```
16:35 $ java -jar amm/target/scala-2.12/ammonite-version-SNAPSHOT-2.12.2  ~/test.sc  --silent -- arg1 arg2
List(arg1, arg2)
 ~/src/oss/Ammonite [master ↑·1|✚ 1…61]
16:35 $ java -jar amm/target/scala-2.12/ammonite-version-SNAPSHOT-2.12.2  ~/test.sc  --silent arg1 arg2
List(arg1, arg2)
 ~/src/oss/Ammonite [master ↑·1|✚ 1…61]
16:35 $ java -jar amm/target/scala-2.12/ammonite-version-SNAPSHOT-2.12.2  ~/test.sc  --silent arg1 -- arg2
Error: You may not provide passThrough args before and after the -- delimiter
 ✘-1 ~/src/oss/Ammonite [master ↑·1|✚ 1…61]
16:36 $
```

**Comment by timcharper** [comment]
Not sure if I caused these failures. `ammonite.session.ProjectTests` passes for me, locally:

```
[info] ------------------Starting Suite ammonite.session.ProjectTests------------------
[info] ammonite.session.ProjectTests.load.ivy.standalone		Success
[info] ammonite.session.ProjectTests.load.ivy.akkahttp		Success
[info] ammonite.session.ProjectTests.load.ivy.resolvers		Success
[info] ammonite.session.ProjectTests.load.ivy		Success
[info] ammonite.session.ProjectTests.load.code		Success
[info] ammonite.session.ProjectTests.load		Success
[info] ammonite.session.ProjectTests.shapeless		Success
[info] ammonite.session.ProjectTests.scalaz		Success
[info] ammonite.session.ProjectTests.cats		Success
[info] ammonite.session.ProjectTests.guava		Success
[info] ammonite.session.ProjectTests.resources		Success
[info] ammonite.session.ProjectTests.scalaparse		Success
[info] ammonite.session.ProjectTests.finagle		Success
[info] ammonite.session.ProjectTests.spire		Success
[info] ammonite.session.ProjectTests.pegdown		Success
[info] ammonite.session.ProjectTests		Success
[info] -----------------------------------Results-----------------------------------
[info] ammonite.session.ProjectTests		Success
[info]     load		Success
[info]         ivy		Success
[info]             standalone		Success
[info]             akkahttp		Success
[info]             resolvers		Success
[info]         code		Success
[info]     shapeless		Success
[info]     scalaz		Success
[info]     cats		Success
[info]     guava		Success
[info]     resources		Success
[info]     scalaparse		Success
[info]     finagle		Success
[info]     spire		Success
[info]     pegdown		Success
[info]
[info] Tests: 16
[info] Passed: 16
[info] Failed: 0
```

File: amm/src/main/scala/ammonite/Main.scala [hunk 1]
```diff
@@ -231,33 +228,51 @@ object Main{
...
+        (args0.slice(0, idx), args0.drop(idx + 1))
     }
 
-    val scriptArgs = ammonite.main.Scripts.groupArgs(flatScriptArgs)
 
     for(c <- replParser.parse(ammoniteArgs, Main())) {
+      val scriptArgs: Seq[(String, Option[String])] =
+        if (flatScriptArgs.nonEmpty && passThroughArgs.nonEmpty) {
+          Console.err.println(
+            "Error: You may not provide passThrough args before and after the -- delimiter")
```
Comment (mgedigian) [inline amm/src/main/scala/ammonite/Main.scala]: "passThrough args" is terminology from the implementation, not user-facing (in usage help or docs)

Comment (timcharper) [inline amm/src/main/scala/ammonite/Main.scala]: Yeah, I thought about this after I'd pushed. Updated.

**Comment by lihaoyi** [comment]
This was changed intentionally in http://www.lihaoyi.com/Ammonite/#0.9.0. The behaviour is documented [here](http://www.lihaoyi.com/Ammonite/#AmmoniteArgumentsinScripts).

A solution is to use the 

```
#!/bin/bash
exec amm "$0" --predef 'println("welcome!")' -- "$@"
!#
```

shebang, as is now (since I made the change) mentioned in the doc-site.

From what I understand, a shebang such as `#!/usr/bin/env amm -s` was never really portable and [doesn't work on Linux](https://stackoverflow.com/questions/4303128/how-to-use-multiple-arguments-with-a-shebang-i-e). You'd have to do the multi-line shebang above anyway if you want to run your scripts on e.g. a server

I'm a bit surprised the tests pass, I guess we don't have much test coverage of the `main` code, but this breaks the following use case:

```scala
lihaoyi Ammonite$ cat foo.sc
@main def main(args: String*) = args

lihaoyi Ammonite$ amm foo.sc --bar # 0.9.x
ArrayBuffer("--bar")

lihaoyi Ammonite$ amm/target/amm foo.sc --bar # PR #619
Error: Unknown option --bar
Try --help for more information.

lihaoyi Ammonite$ amm/target/amm foo.sc -bar # PR #619
Error: Unknown option -bar
Try --help for more information.
```

Which was important for people who wanted to properly "pass through" all arguments for their own custom processing, in cases such as #423, #479, #538.

I'd be curious if you could make this work while maintaining the behaviour shown above, and while maintaining understandability ("always put script first" is relatively straightforward...). Otherwise, you could just swap over to the `exec`-style shebangs and call it a day

**Comment by lihaoyi** [comment]
With some thought, I'm pretty sure we could come up with an implementation that satisfies those requirements, but I don't think we'd be able to use Scopt to do it. We'd need to loop over and deal with the arguments ourselves (as we are already doing in other parts of Ammonite, e.g. the `@main` method router)

**Comment by timcharper** [comment]
@lihaoyi thank you for clarifying! I am ashamed I didn't dig deeper to read the documentation; I simply learned of the behavior change by it breaking when I updated to 0.9.x, finding many of my scripts had broken, and had only read the `--help` screen from Ammonite.

The behavior you describe matches my experience. What was the guiding principle when arriving on that implementation? It seems inconsistent with most interpreter implementations that I've checked.

I have a ruby script, python script, and bash script which simply output their arguments on launch. The following is the output:

```
$ python -d output-args.py -- a b c
['--', 'a', 'b', 'c']

$ python output-args.py -d -- a b c
['-d', '--', 'a', 'b', 'c']

$ python -d -- output-args.py a b c
['a', 'b', 'c']

$ python -d output-args.py a b c
['a', 'b', 'c']

--------

$ ruby -v output-args.rb -- a b c
ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin16]
["--", "a", "b", "c"]

$ ruby output-args.rb -v -- a b c
["-v", "--", "a", "b", "c"]

$ ruby -v -- output-args.rb a b c
ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin16]
["a", "b", "c"]

$ ruby -v output-args.rb a b c
ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin16]
["a", "b", "c"]

--------

$ bash -x output-args.sh -- a b c
+ echo -- a b c
-- a b c

$ bash output-args.sh -x -- a b c
-x -- a b c

$ bash -x -- output-args.sh a b c
+ echo a b c
a b c

$ bash -x output-args.sh a b c
+ echo a b c
a b c
```

It would be possible for me to modify this pull request to support the following cases:

```
amm --predef 'println("welcome!")' script.sc scriptarg1 scriptarg2
amm --predef 'println("welcome!")' -- script.sc scriptarg1 scriptarg2
amm --predef 'println("welcome!")' -- script.sc scriptarg1 scriptarg2

# these do not match convention of bash/ruby/python/etc. but can be supported for backwards compat at the cost of complexity

amm script.sc --predef 'println("welcome!")' -- scriptarg1 scriptarg2
amm --predef 'println("welcome!")' script.sc -- scriptarg1 scriptarg2
```

Would such a change be welcome by you? What is your preference?

**Comment by lihaoyi** [comment]
The guiding principle is "I want to be able to pass in unchanged script args to the script, but I'm too lazy to re-implement Scopt to allow it to properly forward trailing arguments after the script name" 😛

**Comment by lihaoyi** [comment]
I think I would have very much preferred "args coming before script go to amm, args after to go script", but couldn't figure out how to implement that easily; if you want to contribute an implementation I'd take it

**Comment by lihaoyi** [comment]
We have existing code which does the "walk arguments one by one and figure out what to do with them" in here:

- https://github.com/lihaoyi/Ammonite/blob/master/amm/src/main/scala/ammonite/main/Scripts.scala#L19
- https://github.com/lihaoyi/Ammonite/blob/master/amm/src/main/scala/ammonite/main/Router.scala#L54

Would be great if we could re-use the code rather than duplicate it

**Comment by lihaoyi** [comment]
Fixed by https://github.com/lihaoyi/Ammonite/commit/24c424495d9ed012a065dd6adbc17538432a63d5

**Comment by timcharper** [comment]
Awesome, this is much better <3 thanks @lihaoyi !!

**Pull request closed without merging**